relay-server/src/metrics_extraction/metrics_summary.rs

```diff
@@ -0,0 +1,511 @@
+pub struct MetricsSummarySpec {
+    buckets: BTreeMap<MetricsSummaryBucketKey, MetricsSummaryBucketValue>,
+}
+
+impl MetricsSummarySpec {
+    fn new() -> MetricsSummarySpec {
+        MetricsSummarySpec {
+            buckets: BTreeMap::new(),
+        }
+    }
+
+    /// Merges into the [`MetricsSummarySpec`] a slice of [`Bucket`]s.
+    fn from_buckets<'a>(buckets: impl Iterator<Item = &'a Bucket>) -> MetricsSummarySpec {
+        let mut metrics_summary_spec = MetricsSummarySpec::new();
+
+        for bucket in buckets {
+            metrics_summary_spec.merge_bucket(bucket);
+        }
+
+        metrics_summary_spec
+    }
+
+    /// Merges a [`Bucket`] into the [`MetricsSummarySpec`].
+    fn merge_bucket(&mut self, bucket: &Bucket) {
+        let key = MetricsSummaryBucketKey {
+            metric_name: bucket.name.clone(),
+            tags: bucket.tags.clone(),
+        };
+
+        let value = MetricsSummaryBucketValue::from_bucket_value(&bucket.value);
+
+        self.merge(key, value);
+    }
+
+    /// Merges a [`MetricsSummary`] into the [`MetricsSummarySpec`].
+    fn merge_metrics_summary(&mut self, metrics_summary: &MetricsSummary) {
+        for (metric, summary_per_tag) in metrics_summary.0.iter() {
+            let Some(summary_per_tag) = summary_per_tag.value() else {
+                continue;
+            };
+
+            for summary in summary_per_tag {
+                let Some(summary) = summary.value() else {
+                    continue;
+                };
+
+                let mut tags = BTreeMap::new();
+                if let Some(inner_tags) = summary.tags.value() {
+                    for (tag_key, tag_value) in inner_tags.iter() {
+                        let Some(tag_value) = tag_value.value() else {
+                            continue;
+                        };
+
+                        tags.insert(tag_key.clone(), tag_value.clone());
+                    }
+                }
+
+                let key = MetricsSummaryBucketKey {
+                    metric_name: MetricName::from(metric.clone()),
+                    tags,
+                };
+
+                let value = MetricsSummaryBucketValue::from_metric_summary(summary);
+
+                self.merge(key, value);
+            }
+        }
+    }
+
+    /// Merges a [`MetricsSummaryBucketKey`] and [`MetricsSummaryBucketValue`] in the spec.
+    fn merge(&mut self, key: MetricsSummaryBucketKey, value: MetricsSummaryBucketValue) {
+        match self.buckets.entry(key) {
+            Entry::Occupied(mut entry) => {
+                entry.get_mut().merge(value);
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(value);
+            }
+        }
+    }
+
+    /// Applies the [`MetricsSummarySpec`] on a receiving [`Annotated<MetricsSummary>`].
+    ///
+    /// In case the [`MetricsSummarySpec`] is empty, no mutation will take place.
```

Comment:
This is an implementation detail, not necessary for the docs.

Applying an no summaries to some summaries should always yield the same original summaries regardless.